File: examples/quick-git/internal/ui/graph.go
package ui

import (
	"fmt"
	"math"
	"strings"

	gui "github.com/egoist/quickgui/go/ui"
	"quickgui.example/quick-git/internal/git"
)

const (
	historyRowHeight = 26
	graphLaneWidth   = 14
	graphInset       = 6
)

type graphLayer struct {
	Color  int
	Source string
}

func graphLaneX(lane int) float64 {
	return graphInset + float64(lane)*graphLaneWidth + graphLaneWidth/2
}

// Each layer is an SVG mask tinted with one lane color. Paths reach the row's
// exact edges so the virtualized rows form a continuous graph.
func paintGraphRow(row git.GraphRow, width, height float64) []graphLayer {
	middle := height / 2
	radius := math.Min(graphLaneWidth/2, math.Max(1, middle-1))
	segments := map[int][]string{}
	var colors []int
	add := func(color int, path string) {
		if _, ok := segments[color]; !ok {
			colors = append(colors, color)
		}
		segments[color] = append(segments[color], path)
	}
	x := graphLaneX(row.Lane)
	for _, lane := range row.Passing {
		add(lane.Color, fmt.Sprintf("M%g 0V%g", graphLaneX(lane.Lane), height))
	}
	if row.Incoming {
		add(row.Color, fmt.Sprintf("M%g 0V%g", x, middle))
	}
	for _, join := range row.Joins {
		from := graphLaneX(join.FromLane)
		if from == x {
			add(join.Color, fmt.Sprintf("M%g 0V%g", x, middle))
			continue
		}
		direction, sweep := 1.0, 0
		if from > x {
			direction, sweep = -1, 1
		}
		add(join.Color, fmt.Sprintf("M%g 0V%gA%g %g 0 0 %d %g %gH%g", from, middle-radius, radius, radius, sweep, from+direction*radius, middle, x))
	}
	for _, edge := range row.Edges {
		to := graphLaneX(edge.ToLane)
		if to == x {
			add(edge.Color, fmt.Sprintf("M%g %gV%g", x, middle, height))
			continue
		}
		direction, sweep := -1.0, 0
		if to > x {
			direction, sweep = 1, 1
		}
		add(edge.Color, fmt.Sprintf("M%g %gH%gA%g %g 0 0 %d %g %gV%g", x, middle, to-direction*radius, radius, radius, sweep, to, middle+radius, height))
	}
	if _, ok := segments[row.Color]; !ok {
		add(row.Color, "")
	}
	layers := make([]graphLayer, 0, len(colors))
	for _, color := range colors {
		var body strings.Builder
		if path := strings.Join(segments[color], ""); path != "" {
			fmt.Fprintf(&body, `<path d="%s" fill="none" stroke="#000" stroke-width="2"/>`, path)
		}
		if color == row.Color {
			fmt.Fprintf(&body, `<circle cx="%g" cy="%g" r="4" fill="#000"/>`, x, middle)
		}
		layers = append(layers, graphLayer{Color: color, Source: fmt.Sprintf(`<svg xmlns="http://www.w3.org/2000/svg" width="%g" height="%g" viewBox="0 0 %g %g">%s</svg>`, width, height, width, height, body.String())})
	}
	return layers
}

func historyGraph(row func() *git.GraphRow, width func() float64) {
	app := UseApp()
	gui.View(
		gui.Position("relative"),
		gui.Width(width),
		gui.Height(historyRowHeight),
		gui.FlexShrink(0),
		gui.Overflow("hidden"),
		func() {
			gui.KeyedFor(
				func() []graphLayer {
					if current := row(); current != nil {
						return paintGraphRow(*current, width(), historyRowHeight)
					}
					return nil
				},
				func(layer graphLayer) any { return layer.Color },
				func(layer func() graphLayer, _ func() int) {
					gui.SVG(
						gui.Position("absolute"),
						gui.Left(0),
						gui.Top(0),
						gui.Width(width),
						gui.Height(historyRowHeight),
						gui.Color(func() string {
							palette := app.Theme().Graph
							return palette[layer().Color%len(palette)]
						}),
						gui.Value(func() string { return layer().Source }),
					)
				},
				nil,
			)
		},
	)
}

File: examples/quick-git-typescript/ui/graph.ts
/**
 * Painting of one commit-graph row as SVG.
 *
 * The framework rasterizes an `Svg` element into a one-channel mask tinted by its `color`, so a
 * row is painted as one document per lane color: continuous vertical lines for lanes that pass
 * through, the node's own line coming in from above, quarter-circle elbows where a line leaves
 * the node for another lane or arrives from one, and a filled disc for the commit. Adjacent rows
 * paint every lane at the same x and up to their shared edge, so the lines read as continuous.
 */

import type { GraphRow } from "../git/log.ts";

export const LANE_WIDTH = 14;
/** Horizontal inset before the first lane. */
export const GRAPH_INSET = 6;
export const LINE_WIDTH = 2;
export const NODE_RADIUS = 4;

/** Center x of a lane inside the graph cell. */
export function laneX(lane: number): number {
  return GRAPH_INSET + lane * LANE_WIDTH + LANE_WIDTH / 2;
}

/** One SVG document painting every segment of one color in a row. */
export interface GraphLayer {
  /** Lane color index into the theme palette. */
  color: number;
  source: string;
}

/**
 * SVG documents that paint `row` in a cell of `width` × `height` logical pixels, one per color.
 *
 * Elbows are quarter circles whose radius is half a lane, so a line leaving the node runs level
 * for the rest of the lane and turns down exactly at its new lane; a line arriving from above does
 * the same in mirror. Straight segments use butt caps so a lane continues seamlessly into the
 * neighbouring rows.
 */
export function paintGraphRow(row: GraphRow, width: number, height: number): GraphLayer[] {
  const middle = height / 2;
  const radius = Math.min(LANE_WIDTH / 2, Math.max(1, middle - 1));
  const segments = new Map<number, string[]>();
  const add = (color: number, path: string): void => {
    const list = segments.get(color);
    if (list) list.push(path);
    else segments.set(color, [path]);
  };
  const nodeX = laneX(row.lane);
  for (const lane of row.passing) add(lane.color, `M${laneX(lane.lane)} 0V${height}`);
  if (row.incoming) add(row.color, `M${nodeX} 0V${middle}`);
  for (const join of row.joins) {
    const fromX = laneX(join.fromLane);
    if (fromX === nodeX) {
      add(join.color, `M${nodeX} 0V${middle}`);
      continue;
    }
    // Travel direction toward the node decides which way the elbow turns: arriving from the
    // right turns clockwise on screen (sweep 1), arriving from the left counterclockwise.
    const direction = fromX > nodeX ? -1 : 1;
    const sweep = direction < 0 ? 1 : 0;
    add(
      join.color,
      `M${fromX} 0V${middle - radius}A${radius} ${radius} 0 0 ${sweep} ${fromX + direction * radius} ${middle}H${nodeX}`,
    );
  }
  for (const edge of row.edges) {
    const toX = laneX(edge.toLane);
    if (toX === nodeX) {
      add(edge.color, `M${nodeX} ${middle}V${height}`);
      continue;
    }
    const direction = toX > nodeX ? 1 : -1;
    const sweep = direction > 0 ? 1 : 0;
    add(
      edge.color,
      `M${nodeX} ${middle}H${toX - direction * radius}A${radius} ${radius} 0 0 ${sweep} ${toX} ${middle + radius}V${height}`,
    );
  }
  if (!segments.has(row.color)) segments.set(row.color, []);

  const layers: GraphLayer[] = [];
  for (const [color, paths] of segments) {
    const stroke =
      paths.length > 0
        ? `<path d="${paths.join("")}" fill="none" stroke="#000" stroke-width="${LINE_WIDTH}"/>`
        : "";
    const node =
      color === row.color
        ? `<circle cx="${nodeX}" cy="${middle}" r="${NODE_RADIUS}" fill="#000"/>`
        : "";
    layers.push({
      color,
      source: `<svg xmlns="http://www.w3.org/2000/svg" width="${width}" height="${height}" viewBox="0 0 ${width} ${height}">${stroke}${node}</svg>`,
    });
  }
  return layers;
}
